lib/ttfunk/subset_collection.rb

```diff
@@ -17,12 +17,16 @@ def [](subset)
     def use(characters)
       characters.each do |char|
         covered = false
-        @subsets.each_with_index do |subset, _i|
```

Comment:
Why are we even using `each_with_index` if we never use the index anywhere? The `while` loop is fine, but can we get the same allocation wins with a plain 'ol `each`?

Reply:
As far as I know the reason for the allocation is the block itself. So if we would use `each` instead of `each_with_index` there would still be an allocation.

Furthermore, there is an non-negligible performance hit when using a block instead of a plain `while` loop due to the block invocation. So apart from saving the allocation the `while` loop is also faster which is especially visible in such an often called inner loop.

Reply:
That's super interesting, I wonder why blocks cause an allocation? I do vaguely remember hearing something about that at a Railsconf one time though.

Reply:
Yes, there was a talk some time ago that went through (nearly) all the Enumerable methods to show the time/space complexity - I also don't remember its title.  However, there is a great talk by Jeremy Evans about optimizations in Roda/Sequel: http://code.jeremyevans.net/presentations/rubykaigi2019/index.html#83 (this links directly to the "avoid proc allocation slide").